packages/publisher/src/publish-packages.ts

```diff
@@ -187,6 +192,85 @@ export default async function publishPackages(
     cacheDirPath
   );
 
+  // Loop over the @types packages in npm and mark any that no longer
+  // exist in HEAD as deprecated.
+  let from = 0;
+  let objects;
+  do {
+    const opts = {
+      limit: 250,
+      from
+    };
+    objects = await search("@types", opts);
+    for (const { name: fullNpmName } of objects) {
+      const name = fullNpmName.slice("@types/".length);
+      // If they don't exist in the types directory or in
+      // notNeededPackages.json then mark them deprecated. Reference the
+      // commit/pull request that removed them.
+      if (!allPackages.tryGetLatestVersion(name) && !allPackages.getNotNeededPackage(name)) {
+        log(`Deprecating ${name}`);
+        const {
+          repository: {
+            ref: {
+              target: {
+                history: {
+                  nodes: [commit]
+                }
+              }
+            }
+          }
+        } = await graphql(
+          `
+            query($path: String!) {
+              repository(name: "DefinitelyTyped", owner: "DefinitelyTyped") {
+                ref(qualifiedName: "master") {
+                  target {
+                    ... on Commit {
+                      history(first: 1, path: $path) {
+                        nodes {
+                          associatedPullRequests(first: 1) {
+                            nodes {
+                              url
+                            }
+                          }
+                          messageHeadline
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          `,
+          {
+            headers: { authorization: `token ${githubAccessToken}` },
+            path: `types/${name}`
+          }
+        );
+        let deprecatedMessage;
+        if (commit) {
+          const {
+            associatedPullRequests: {
+              nodes: [pullRequest]
+            },
+            messageHeadline
+          } = commit;
+          deprecatedMessage = messageHeadline;
+          if (pullRequest) {
+            deprecatedMessage += ` (${pullRequest.url})`;
+          }
+        }
+        if (dry) {
+          log(`(dry) Skip deprecate removed package ${fullNpmName}`);
+        } else {
+          log(`Deprecating ${fullNpmName} with message: ${deprecatedMessage}`);
+          await client.deprecate(fullNpmName, "", deprecatedMessage);
+        }
+      }
+    }
+    from += objects.length;
+  } while (objects.length >= 250 && from <= 5000);
```

Comment:
what is 5000? I thought there were close to 7000 `@types` packages.

Reply:
Yes, so unfortunately the best I've been able to do so far is loop over the first 5250 `@types` packages ... I hope it can be built upon and that it's useful in the meantime ...

The only way I've found to loop over npm packages is the [registry API](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md#get-v1search) and it clamps `limit` at 250 and `from` at 5000. Without the `from <= 5000` condition it just repeats results 5000-5249 forever. The condition can be > 5000, it just repeats unnecessarily, and < 5000 gets fewer distinct results. So I think it's necessary and 5000 is optimal.

I hunted around for alternatives:
- I tried the [npms API](https://api-docs.npms.io/#api-Search) but the results I got were out of date. And it has a max of 5000 anyway.
- I looked for a dump of registry package names or a [public dataset](https://azure.microsoft.com/en-ca/services/open-datasets/catalog/?q=npm). No luck.
- I considered changing the sort parameters to e.g. get the last 5000 results plus the first 5000, which would cover all the `@types` packages of today, but think it's inelegant and ineffective: The registry API supports weighting results differently but not flipping the order entirely.

So [libnpmsearch](https://github.com/npm/libnpmsearch) is imperfect but it's fast, officially supported and has non-zero utility (I think). If it runs periodically I hope it'll catch results while they're in the first 5000. And hopefully one day the limit will increase, or an improved API will debut, or someone will think of something else?